content/posts/redis-assisted-client-side-caching-in-python.md

```diff
@@ -0,0 +1,213 @@
+
+The requirements from the cache component are basic, so I was perfectly happy adapting the [LRU cache example in Python's `OrederedDict` documentation](https://docs.python.org/3/library/collections.html#ordereddict-examples-and-recipes).
+
+## Making a regular connection into a cached one
+For this experiment, I chose to implement caching for a single Redis command, namely [`GET`](https://redis.io/commands/get). The premise is to make the client use the _read through_ pattern: that is, to attempt a read from the local cache and defer to Redis in case of a miss. Subclassing the `Redis` client and overriding its `get()` method gives us the following:
+
+```python
+class Redis(redis.Redis):
+    def __init__(self, manager, *args, **kwargs):
+        super().__init__(self, *args, **kwargs)
+        self._manager = manager
+        self._client_id = super().client_id()
+        self.execute_command('CLIENT', 'TRACKING', 'ON',
+            'redirect', self._manager.client_id)
+
+    def get(self, name):
+        try:
+            value = self._manager.cache[name]
+        except KeyError:
+            value = super().get(name)
+            self._manager.cache[name] = value
+        return value
+```
+
+The new `Redis` class initialization isn’t fancy. It begins by calling its base class' `__init__()` and setting a couple of properties. It ends with a call to Redis' `CLIENT TRACKING` command, with which it enables the server's assistance for the connection.
+
+The class' `get()` method is where the magic happens. We try to read the key's value, by name, from the cache that's available through the connection's manager. In case of `KeyError`, or a cache miss, we revert to the base class' `get()` to fetch the data and store it in the cache.
+
+## Tracking keys in the client
+Once the Redis client has opted-in to being tracked, the server maintains a record of keys that the client had read from. Instead of tracking every individual key, Redis uses a hashing function on the keys' names to assign them to slots. Specifically, it uses the 24 least-significant bits of the key name's CRC64 digest, resulting in roughly 16 million possible slots.
+
+This reduces the server resources required to track multiple keys for multiple clients. The invalidation messages sent by Redis, therefore, consist of _slots_ that need to be invalidated rather than key names. It is up to the client to infer the relevant key names that need to be removed from its cache given the slot.
+
+That means the client needs to employ the same hashing function to track how the keys in the local cache map to slots. That lets us perform slot-based invalidation of the client's cache when an invalidation notification arrives. For that, we'll use the `add()` and `discard()` methods when keys are added and discarded from the local cache, respectively.
+
+```python
+    def slot(key):
+        ''' Returns the slot for a key '''
+        crc = crc64(key)
+        crc %= 16777216
```

Comment:
Writing `0x1000000` would be more readable.

Reply:
Right. I'll do `&= 0xffffff` instead though.